open-api/rest-catalog-open-api.yaml

```diff
@@ -3103,6 +3103,95 @@ components:
         uuid:
           type: string
 
+    ADLSCredential:
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/Credential'
+      required:
+        - type
+        - sas-token
+        - expires-at-ms
+      properties:
+        type:
+          type: string
+          enum: [ "adls" ]
+        prefix:
+          type: string
+          description: Indicates a storage location prefix where the credential is relevant. Clients should choose the most
+            specific prefix if several credentials of the same type are available.
+        sas-token:
+          type: string
+        expires-at-ms:
+          type: integer
+          format: int64
+          description: The epoch millis since 1970-01-01T00:00:00Z at which the given token expires
+
+
+    GCSCredential:
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/Credential'
+      required:
+        - type
+        - token
+        - expires-at-ms
+      properties:
+        type:
+          type: string
+          enum: [ "gcs" ]
+        prefix:
+          type: string
+          description: Indicates a storage location prefix where the credential is relevant. Clients should choose the most
+            specific prefix if several credentials of the same type are available.
+        token:
+          type: string
+        expires-at-ms:
+          type: integer
+          format: int64
+          description: The epoch millis since 1970-01-01T00:00:00Z at which the given token expires
+
+    S3Credential:
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/Credential'
+      required:
+        - type
+        - access-key-id
+        - secret-access-key
+        - session-token
```

Comment:
do we need to require `session-token` and `expires-at-ms`? This forces the credentials to be temporary, which is probably the case most of the time, but I don't think we need to exclude the case where people want to do differently. I think some systems using S3 as a proxy might be able to appreciate this.

Reply:
Yes we require those specifically to make the credentials temporary and not accidentally leak non-expiring credentials (see also https://docs.google.com/document/d/1lySd_5hMZNtISLKsOvAq7xiNzdXU6TAoHF_yrOXWQvM/edit#heading=h.hs6r9d26w1y2 where this is mentioned)

Reply:
That makes sense. In that case I think we can keep `expires-at-ms` required. But what about `session-token`? I think not all storage systems that implements the S3 API supports that.

I know we want to explore defining "credentials" as a map, but assuming we go with this current approach, I feel it is better to at least make that optional.

Reply:
I agree with @jackye1995 that `session-token` is not part of the S3 API, and users might not have IAM or similar setups in their environment. I understand the motivation for requiring `session-token` to exist, but I'm concerned it could be too restrictive and prevent users from implementing the rest spec.

Reply:
@Xuanwo the PR has been reworked to not include properties specific to storage providers, so this shouldn't be an issue anymore. See also https://github.com/apache/iceberg/pull/10722#discussion_r1795062381

Reply:
Hi, @nastra. Sorry for my outdated comment, and thank you for this change. This should work.